### PR TITLE
Add DataflowStartFlexTemplateOperator

### PR DESCRIPTION
**Pull request by mik-laj**
---
Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Target Github ISSUE in description if exists
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by mik-laj** [comment]
@jaketf Can I ask for review? I know that you are also interested in integration with Dataflow. Will this solve your clients' problems fully? If they use it, they will no longer have to use KubernetesPodOperator.

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 1]
```diff
@@ -93,15 +93,30 @@ def inner_wrapper(self: "DataflowHook", *args, **kwargs) -> RT:
 class DataflowJobStatus:
     """
     Helper class with Dataflow job statuses.
+    Reference: https://cloud.google.com/dataflow/docs/reference/rest/v1b3/projects.jobs#Job.JobState
```
Comment (aaltay) [inline airflow/providers/google/cloud/hooks/dataflow.py:99]: This file is also modified in https://github.com/apache/airflow/pull/8553. I am assuming it is the same changes, skipping this file for the review.

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 2]
```diff
@@ -406,6 +406,71 @@ def on_kill(self) -> None:
...
+            project_id=self.project_id,
+            on_new_job_id_callback=set_current_job_id,
+        )
+
+        return job
+
+    def on_kill(self) -> None:
+        self.log.info("On kill.")
+        if self.job_id:
+            self.hook.cancel_job(job_id=self.job_id, project_id=self.project_id)
```
Comment (aaltay) [inline airflow/providers/google/cloud/operators/dataflow.py:517]: Do you need to call this if job is no longer running?

File: tests/providers/google/cloud/operators/test_dataflow_system.py [hunk 3]
```diff
@@ -26,4 +37,198 @@
...
+    def setUp(self) -> None:
+        # Create a Cloud Storage bucket
+        self.execute_cmd(["gsutil", "mb", f"gs://{GCS_FLEX_TEMPLATE_BUCKET_NAME}"])
+
+        # Build image with pipeline
+        with NamedTemporaryFile() as f:
+            f.write(
+                textwrap.dedent(
+                    """\
+                    steps:
```
Comment (aaltay) [inline tests/providers/google/cloud/operators/test_dataflow_system.py]: What is being built in this cloud image pipeline?

File: tests/providers/google/cloud/operators/test_dataflow_system.py [hunk 4]
```diff
@@ -15,9 +15,20 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+import os
```
Comment (aaltay) [inline tests/providers/google/cloud/operators/test_dataflow_system.py:19]: This file reads more like an end2end test of various dataflow system things. Should it be a different PR?

File: airflow/providers/google/cloud/example_dags/example_dataflow_flex_template.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+with models.DAG(
+    dag_id="example_gcp_dataflow_flex_template_java",
+    default_args={
+        "start_date": days_ago(1),
+    },
+    schedule_interval=None,  # Override to match your needs
+) as dag_flex_template:
+    start_flex_template = DataflowStartFlexTemplateOperator(
+        task_id="start_flex_template_java",
```
Comment (aaltay) [inline airflow/providers/google/cloud/example_dags/example_dataflow_flex_template.py]: maybe drop java from the name? (Given a template structure, what language was used to write it matters less.)

**Review by aaltay — COMMENTED** [review]
/cc: @azurezyq

File: tests/providers/google/cloud/operators/test_dataflow_system.py (hunk 3 above)
Comment (mik-laj) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: I am building an image from your repository.
https://github.com/GoogleCloudPlatform/java-docs-samples/tree/954553c/dataflow/flex-templates/streaming_beam_sql

Comment (aaltay) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: Ack. Does it required this many containers in the image?

I am not very familiar. Is this a VM image with containers?

File: tests/providers/google/cloud/operators/test_dataflow_system.py [hunk 6]
```diff
@@ -26,4 +37,198 @@
...
+                    - name: alpine
+                      entrypoint: /bin/sh
+                      dir: 'repo_dir/$_EXAMPLE_SUBDIR'
+                    - name: maven
+                      args: ["mvn", "clean", "package"]
+                      dir: 'repo_dir/$_EXAMPLE_SUBDIR'
+                    - name: alpine
+                      entrypoint: "/bin/sh"
+                      args: ["-c", "ls -lh target/*.jar"]
+                      dir: 'repo_dir/$_EXAMPLE_SUBDIR'
```
Comment (mik-laj) [inline tests/providers/google/cloud/operators/test_dataflow_system.py]:
```suggestion
```

File: tests/providers/google/cloud/operators/test_dataflow_system.py [hunk 7]
```diff
@@ -26,4 +37,198 @@
...
+                    """\
+                    steps:
+                    - name: gcr.io/cloud-builders/git
+                      args: ['clone', '$_EXAMPLE_REPO', 'repo_dir']
+                    - name: gcr.io/cloud-builders/git
+                      args: ['checkout', '$_EXAMPLE_COMMIT']
+                      dir: 'repo_dir'
+                    - name: alpine
+                      entrypoint: /bin/sh
+                      dir: 'repo_dir/$_EXAMPLE_SUBDIR'
```
Comment (mik-laj) [inline tests/providers/google/cloud/operators/test_dataflow_system.py]:
```suggestion
```

File: tests/providers/google/cloud/operators/test_dataflow_system.py (hunk 4 above)
Comment (mik-laj) [reply tests/providers/google/cloud/operators/test_dataflow_system.py:19]: We always try to write system tests and integrations simultaneously to ensure the best integration reliability. Thanks to this, every person on my team can easily test every integration.

File: tests/providers/google/cloud/operators/test_dataflow_system.py (hunk 3 above)
Comment (mik-laj) [reply tests/providers/google/cloud/operators/test_dataflow_system.py:69]: The Cloud Build service will provide a new virtual machine for each build. Each build contains many steps. Each step is described by a container image. I personally would not worry about a large number of images, because after the build is completed, the virtual machine is deleted.

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 2 above)
Comment (mik-laj) [reply airflow/providers/google/cloud/operators/dataflow.py:517]: Good point. I will change it

File: tests/providers/google/cloud/operators/test_dataflow_system.py (hunk 3 above)
Comment (mik-laj) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: I removed some build steps. 
https://github.com/apache/airflow/pull/8550/commits/aa29389cac91f29f5f37b268fa61c0a3eaabd663

In order to build this Docker image I used Google Cloud Build. The Cloud Build service will provide a new virtual machine for each build. Each build contains many steps. Each step is described by a Docker image. As a result of the build, a new Docker image is built.

I personally would not worry about a large number of images, because after the build is completed, the virtual machine is deleted. It does not run on the local machine.

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 8]
```diff
@@ -583,6 +623,49 @@ def start_template_dataflow(
...
+        self,
+        body: Dict,
+        location: str,
+        project_id: str,
+        on_new_job_id_callback: Optional[Callable[[str], None]] = None
+    ):
+        """
+        Starts flex templates with the Dataflow  pipeline.
+
+        :param body: The request body
```
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/dataflow.py]: nit: use more descriptive name `launch_flex_template_parameters`

If we are not going to construct this body here and document the parameters then this should include link tot he API spec for this model.

Unfortunately it looks like we mention this end point [here in flex templates docs](https://cloud.google.com/dataflow/docs/guides/templates/using-flex-templates) but not here in the [REST API docs](https://cloud.google.com/dataflow/docs/reference/rest).

I've opened an internal docs bug on this.

~In the mean time I'm pretty sure this body is just the same models as the "old" templates endpoint uses [LaunchTemplateParameters](https://cloud.google.com/dataflow/docs/reference/rest/v1b3/LaunchTemplateParameters).
@aaltay can you confirm?~ 

Looks like LaunchTemplateParameters is missing the `container_spec_gcs_path` key which is necessary for this endpoint.
Why is this key snake_case in the API while all others (e.g. jobName) are camelCase anyway?

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 9]
```diff
@@ -406,6 +406,71 @@ def on_kill(self) -> None:
             self.hook.cancel_job(job_id=self.job_id, project_id=self.project_id)
 
 
+class DataflowStartFlexTemplateOperator(BaseOperator):
+    """
+    Starts flex templates with the Dataflow  pipeline.
+
+    :param body: The request body
```
Comment (jaketf) [inline airflow/providers/google/cloud/operators/dataflow.py]: see comment on same param in hook.

File: tests/providers/google/cloud/hooks/test_dataflow.py [hunk 10]
```diff
@@ -984,15 +1034,17 @@ def test_dataflow_job_is_job_running_with_no_job(self):
...
-
         get_method = (
             self.mock_dataflow.projects.return_value.
             locations.return_value.
             jobs.return_value.
             get
         )
-        get_method.return_value.execute.return_value = job
+        get_method.return_value.execute.side_effect = [
+            {"id": TEST_JOB_ID, "name": TEST_JOB_NAME, "currentState": DataflowJobStatus.JOB_STATE_RUNNING},
```
Comment (jaketf) [inline tests/providers/google/cloud/hooks/test_dataflow.py]: note from other PR #8553 see how `JobStatus.JOB_STATE` is redundant stutter?
Especially if we allow user to control failed states we should remove the stutter.

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 11]
```diff
@@ -583,6 +623,49 @@ def start_template_dataflow(
...
+            on_new_job_id_callback(job_id)
+
+        jobs_controller = _DataflowJobsController(
+            dataflow=self.get_conn(),
+            project_number=project_id,
+            job_id=job_id,
+            location=location,
+            poll_sleep=self.poll_sleep,
+            num_retries=self.num_retries)
+        jobs_controller.wait_for_done()
```
Comment (jaketf) [inline airflow/providers/google/cloud/hooks/dataflow.py:692]: @mik-laj was reflecting on this in light of the data fusion operator issue.
This is "start" naming confusing.

This method (and Dataflow*Start*FlexTemplateOperator) are called "start" flex template but this appears like you are waiting for the job to complete.

The existing dataflow operators do not have this start word and I think the user expectation is that they poll the job to completion. Otherwise you can't do much useful downstream in the DAG without having some sensor that waits on this job completion.

If we want to support blocking or not blocking I'd suggest having a `wait_for_done` kwarg that defaults to `True` (the expected behavior based on similar operators). This might mean that we need a new method in the controller `wait_for_running` that blocks until the pipeline enters the RUNNING state.

What do you think?

same applies for #8553

Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py:692]: I am trying to imitate the naming conventions that we currently have in this integration.
DataflowTemplatedJobStartOperator
It uses the verb "Start" to describe identical operations.  

I plan to add a blocking and non-blocking mode to all operators in a separate PR, because this requires the development of sensors. For now, I only wanted to focus on one issue.

**Comment by syucream** [comment]
What's the current status? Do you have any progress? I want this patch, Can I help you anything? @mik-laj

**Comment by mik-laj** [comment]
Hiello. 
I had urgent tasks that prevented me from continuing my work. Now I am continuing work on Dataflow, but for now I have to fix problems with system tests.I hope to finish it next week.
Best regards

**Comment by syucream** [comment]
@mik-laj Any update? I created that pull request https://github.com/PolideaInternal/airflow/pull/952 to fix the conflict and some points which were discussed here. Can you see it if it's helpful?

File: tests/providers/google/cloud/operators/test_dataflow_system.py [hunk 12]
```diff
@@ -26,4 +37,191 @@
...
+    def setUp(self) -> None:
+        # Create a Cloud Storage bucket
+        self.execute_cmd(["gsutil", "mb", f"gs://{GCS_FLEX_TEMPLATE_BUCKET_NAME}"])
+
+        # Build image with pipeline
+        with NamedTemporaryFile() as f:
+            f.write(
+                textwrap.dedent(
+                    """\
+                    steps:
```
Comment (ad-m) [inline tests/providers/google/cloud/operators/test_dataflow_system.py]: Have you considered use standard `dict` and use `json.dumps` to generate file content? You don't use any YAML specific syntax and JSON is subset of YAML. I believe that `dict` is easier to maintain long term than a hard-coded text block.

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 2 above)
Comment (TobKed) [reply airflow/providers/google/cloud/operators/dataflow.py:517]: I changed it here https://github.com/apache/airflow/pull/8550/commits/1eac7722208d6f7e7723eee44c676c4f0bc5065a

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 1 above)
Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py:99]: I replied here https://github.com/apache/airflow/pull/8553#discussion_r496513620 and changed here https://github.com/apache/airflow/pull/8550/commits/92858ebbc39c242376304d4a9d9106263aed7369

File: airflow/providers/google/cloud/example_dags/example_dataflow_flex_template.py (hunk 5 above)
Comment (TobKed) [reply airflow/providers/google/cloud/example_dags/example_dataflow_flex_template.py]: I renamed it to `start_flex_template_streaming_beam_sql` here https://github.com/apache/airflow/pull/8550/commits/d788bab82047230829542e9eb8413d16b3565587

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 8 above)
Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py]: I see documentation of the request body is available now:
https://cloud.google.com/dataflow/docs/reference/rest/v1b3/projects.locations.flexTemplates/launch#request-body
Thanks taking care of it @jaketf 
I added links to in hook and operator.

In relation to naming I think snake_case is used as standard for python and camelCase when it is required by external APIs (usually within dictionaries). I am not sure I understand correctly your question. Could you elaborate please?

File: tests/providers/google/cloud/operators/test_dataflow_system.py (hunk 12 above)
Comment (TobKed) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: I think it was used for sake of simplicity. It is easier to take a look on the multiline string which represents quite simple YAML. Am I right @mik-laj ?

Comment (ad-m) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: Is it easier to read a YAML dictionary inserted as text (no formatting in most IDE) than a Python dictionary with correct editor formatting?

Comment (TobKed) [reply tests/providers/google/cloud/operators/test_dataflow_system.py]: You are right, I changed my mind. I think dictionary looks better especially that Cloud Build accept JSON config files

```
cloud_build_config = {
    'steps': [
        {'name': 'gcr.io/cloud-builders/git', 'args': ['clone', "$_EXAMPLE_REPO", "repo_dir"]},
        {
            'name': 'gcr.io/cloud-builders/git',
            'args': ['checkout', '$_EXAMPLE_COMMIT'],
            'dir': 'repo_dir',
        },
        {
            'name': 'maven',
            'args': ['mvn', 'clean', 'package'],
            'dir': 'repo_dir/$_EXAMPLE_SUBDIR',
        },
        {
            'name': 'gcr.io/cloud-builders/docker',
            'args': ['build', '-t', '$_TEMPLATE_IMAGE', '.'],
            'dir': 'repo_dir/$_EXAMPLE_SUBDIR',
        },
    ],
    'images': ['$_TEMPLATE_IMAGE'],
}
```

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 2 above)
Comment (TobKed) [reply airflow/providers/google/cloud/operators/dataflow.py:517]: Improved it here https://github.com/apache/airflow/pull/8550/commits/4b78b27d855cd61d97ffd3d2c2591b34a1a2f226

File: tests/providers/google/cloud/hooks/test_dataflow.py (hunk 10 above)
Comment (TobKed) [reply tests/providers/google/cloud/hooks/test_dataflow.py]: created separate issue for that: https://github.com/apache/airflow/issues/11205

**Comment by TobKed** [comment]
I replied to comments and made requested changes. Recently rebased on the latest master.

cc @mik-laj @syucream @ad-m @aaltay @jaketf

**Review by ad-m — APPROVED** [review]

**Review by aaltay — APPROVED** [review]

**Review by jaketf — APPROVED** [review]

**Review by ad-m — CHANGES_REQUESTED** [review]
Could you rebase?

**Comment by TobKed** [comment]
> Could you rebase?

@ad-m I rebased on the lastest master and squashed commits.

**Comment by TobKed** [comment]
@mik-laj could you remove **dont-merge** label, please?
cc @potiuk @turbaszek

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 13]
```diff
@@ -282,41 +295,70 @@ def wait_for_done(self) -> None:
...
 
         return self._jobs
 
+    def _wait_for_states(self, expected_states: Set[str]):
+        """
+        Waiting for the jobs to reach a certain state.
+        """
+        if not self._jobs:
+            raise ValueError("The _jobs should be set")
+        while True:
```
Comment (turbaszek) [inline airflow/providers/google/cloud/hooks/dataflow.py]: Should we consider implementing a timeout?

Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py]: We have timeout: See: https://airflow.readthedocs.io/en/latest/_api/airflow/models/index.html#airflow.models.BaseOperator -> `execution_timeout`

Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py]: PTAL: https://github.com/apache/airflow/pull/8550/commits/144b63f4ec9835d8c6c57816ab04761b83bee6c2

Since the `cancel` method is executed only by the operator (`on_kill`) I didn't allow user to configure this timeout because I don't think it is worth to add complexity to the user and the code itself. It is added to prevent hanging this `_wait_for_states` forever when `execution_timeout` is not set and provide more meaningful log message if the timeout eventually occur.

I proposed 5 minutes of default timeout.  @aaltay @kamilwu what do you think about this value ?

Comment (aaltay) [reply airflow/providers/google/cloud/hooks/dataflow.py]: I think this is reasonable.

Side note, cancel usually completes quickly. On the other hand drain is the safer way to cancel streaming pipelines and that can take a long time depending on the state of the pipeline.

Comment (turbaszek) [reply airflow/providers/google/cloud/hooks/dataflow.py]: If we use this only in on_kill, should we even bother with waiting? I would assume that a successful API call should be enough for us

Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py]: @mik-laj what do you think about removing waiting? I lean towards @turbaszek opinion

**Review by ad-m — APPROVED** [review]
LGTM

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 13 above)
Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py]: @aaltay I created separated draft PR for handling draining on kill: https://github.com/apache/airflow/pull/11374. WDYT?

Comment (TobKed) [reply airflow/providers/google/cloud/hooks/dataflow.py]: Since it is not strictly related changed and I don't want to block this PR I deleted this part of logic and created separated PR for handling waiting and timeout in job cancel: https://github.com/apache/airflow/pull/11501  (draft because I must add tests).

**Comment by potiuk** [comment]
Hey @TobKed . Can you please rebase this one to the latest master. We fixed (hopefully) a problem with queues of jobs for GitHub actions and I think when you rebase, it should run much faster (more info on devlist shortly).

**Comment by TobKed** [comment]
rebased on the latest master
cc @potiuk @turbaszek

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Thanks @TobKed !

**Comment by ad-m** [comment]
Ja także dziękuje @TobKed za rozwój projektu Airflow!
